### ISSUE TITLE
[$250] [Held requests][HIGH] Hold button on expense preview appears only if open the expense first

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#47241]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.19-0 
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/4845380&group_by=cases:section_id&group_order=asc&group_id=309128
**Email or phone of affected tester (no customers):** gocemate+workspaceapprover157@gmail.com
**Issue reported by:** Applause Internal Team

## Action Performed:
Precondition:
1. Login as the owner of the workspace> Create a workspace
2. Invite the approver and employee
3. Enable "workflows"> On the "Workflow" editor - enable "Add Approvals"> Set the Approver account as the Approver

Steps:
1. Log in as the employee and submit 2 expenses to the workspace
2. Log in as the Approver> Navigate to the employee workspace chat with expenses
3. Right click on expenses to open the menu

## Expected Result:
There should be Hold option when approver open the right click menu from expense

## Actual Result:
Hold button on expense preview appears only if open the transaction thread page first

## Workaround:
Unknown

## Platforms:

- [x] Android: Native
- [x] Android: mWeb Chrome
- [ ] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/3bd3f688-4859-46e5-88fe-00251895e615

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021902104887608709904</li>
        <li>Upwork Job ID: 1902104887608709904</li>
        <li>Last Price Increase: 2025-08-05</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @robertjchen</details>

**Comment by IuliiaHerets** [comment]
We think that this bug might be related to #wave-collect - Release 1

**Comment by cdOut** [comment]
This is definitely related to the changes I've made in the [hold cleanup PR](https://github.com/Expensify/App/pull/45151), please assign me so I can look into fixing this issue.

**Comment by francoisl** [comment]
Sounds good, assigne dto @cdOut and going to demote to non-blocker.

**Comment by trjExpensify** [comment]
I have a feeling this is related to some other issues in [this list](https://expensify.slack.com/archives/C036QM0SLJK/p1723483624957679?thread_ts=1723483611.935549&cid=C036QM0SLJK) and OpenReport. CC: @robertjchen

**Comment by robertjchen** [comment]
It looks like this is still a strictly frontend issue? The `hold` option should be available in the menu without having to open the expense first

**Comment by jliexpensify** [comment]
Assigned you @parasharrajat

**Comment by cdOut** [comment]
Currently waiting on the backend and steps to implement on the frontend suggested by @robertjchen

**Comment by trjExpensify** [comment]
>Currently waiting on the backend and steps to implement on the frontend suggested by @robertjchen 

@robertjchen do we need to assign you to this issue for that backend PR, and is it being worked on somewhere to link here?

**Comment by robertjchen** [comment]
Hm, I'm not sure what backend changes are needed here? My last suggestion here was a frontend approach: https://github.com/Expensify/App/issues/47241#issuecomment-2293798315

Was there a discussion elsewhere that led us to a backend conclusion?

**Comment by robertjchen** [comment]
I'm okay taking this on, but just wanted to make sure we distributed the work and avoid bottlenecking

**Comment by cdOut** [comment]
@robertjchen Well I tried the frontend approach but that meant calling the `openReport` API command which we've settled on discarding, as discussed in [this slack thread](https://swmansion.slack.com/archives/C05LX9D6E07/p1724452830917979?thread_ts=1724312407.953209&cid=C05LX9D6E07). In the end we've decided on adding the `childCanHold` / `childCanUnhold` boolean flags to the report and last week Vit asked for some steps to add support for this change on the frontend side.

**Comment by robertjchen** [comment]
Ah got it- thanks for the context! I'll re-assign to myself and revisit 👀

**Comment by cdOut** [comment]
I've closed the previous PR, @robertjchen please let me know if this moves further and I can help out with the frontend implementation.

**Comment by robertjchen** [comment]
Diving into this task this week now that other issues are out of the way

**Comment by robertjchen** [comment]
Just getting to this! It looks like the `Hold` option on the context menu is determined by the report action preview? https://github.com/Expensify/App/blob/main/src/libs/ReportUtils.ts#L3140

@cdOut Just wanted to check what the frontend will be expecting- does the backend need to push up a new `childCanHold/childCanUbHold ` flag, or can it just update the existing properties on the report action (or report object?).

From what I see, once a particular expense is held, we just need to make an Onyx update to all participants of the report with the new/updated propertie(s).

**Comment by robertjchen** [comment]
Thanks @cdOut . Let me know when you have the details 🙇

**Comment by cdOut** [comment]
I'll list what needs to be added and to what report here in a moment.

**Comment by cdOut** [comment]
@robertjchen What we need to do here is basically add parameters into report which are similar to `hasOutstandingChildRequest` or `hasOutstandingChildTask` that are already implemented, but instead we'll pass `childCanHold / childCanUnhold` into them.

As far as I remember there is already functionality on the backend to get those values based on the child report action so it shouldn't be too difficult to implement and would require very little changes on the frontend as well.

**Comment by robertjchen** [comment]
The BE change has now been deployed! The child reportAction summary will now return the `childCanHold` and `childCanUnhold` state in addition to the existing fields below. Please let me know if you're able to see and use the new fields 👍 

```
childOldestFourAccountIDs
childCommenterCount
childVisibleActionCount
childLastVisibleActionCreated
childType
childReportNotificationPreference
childMoneyRequestCount
childLastMoneyRequestComment
childLastActorAccountID
childType
childLastReceiptTransactionIDs
childRecentReceiptTransactionIDs
childRecentReceiptTransactionIDs
```

**Comment by cdOut** [comment]
Thank you Robert, I'll raise a PR with the newly added params for this one later today.

**Comment by robertjchen** [comment]
Thanks, let me know if you need anything else!

**Comment by cdOut** [comment]
@robertjchen I've tried implementing it with the newly added params, and while it does solve the issue of displaying properly when we can and cannot `hold/unhold`, we are actually missing data required to make an API call required to change the state of the hold.

That is being done through the IOU methods `unholdRequest` and `putOnHold` (This one isn't called directly but it is ran after going through the reason page so we still need data for it). They both require `transactionID` and `reportID` - I tried finding them in the data we have, but the required actions to retrieve them aren't yet loaded if we don't open the request. We'd have to add both of them to the params for me to fix the ability to change hold status.

**Comment by cdOut** [comment]
We could send over one additional param`childHoldData` instead of the two new ones with the following structure:

```
{
        isOnHold: boolean,
        transactionID: string,
        reportID: string,
},
```

Which we'd set to `null/undefined` if it wasn't holdable, `isOnHold` would help us disconcern whether it's currently held or unheld and what method to call on it, and the rest is the data required for me to make the API call through the IOU methods.

**Comment by robertjchen** [comment]
@cdOut Would we need to also include multiple `transactionID`s and `reportID`s if multiple children are on hold? 🤔

**Comment by cdOut** [comment]
> Would we need to also include multiple `transactionID`s and `reportID`s if multiple children are on hold? 🤔

@robertjchen No, I don't think so. what we're implementing here into the context menu is supposed to be an alternative to the promoted action which we have in the report header menu, and in there we can only hold singular requests. Hold isn't currently available for the whole thread, I've attached a video as a reminder (Holding multiple requests at once is available through search, but then we have all the data loaded into onyx so we don't have to include it in this case).

So I'm fairly sure we'd only have to put in a singular `requestID` and a `transactionID` for that specific report preview action.

https://github.com/user-attachments/assets/52a71675-2ac6-4f19-b2ed-9e4c9da30ee6

**Comment by cdOut** [comment]
@robertjchen Do you think the data structure I suggested for it is okay? I already have it implemented on the frontend so I could have it ready before we merge the backend changes.

**Comment by robertjchen** [comment]
From my understanding, instead of returning `childCanHold` and `childCanUnhold`, we are to return a new `childHoldData` object instead, is that correct? 

The problem is that at the point of the logic where we generate those flags on the backend, we get a list of held transactions (which implies there are multiple transactionIDs that need to be returned).

This means that the `childHoldData` object must be returned elsewhere and not at the current location where the `childCanHold` and `childCanUnhold` flags are.

Let me know if that makes sense- could you please link your draft PR so I can see where this is happening? Thanks!

**Comment by robertjchen** [comment]
@cdOut From my understanding, instead of returning `childCanHold` and `childCanUnhold`, we are to return a new `childHoldData` object instead, is that correct? 

The problem is that at the point of the logic where we generate those flags on the backend, we get a list of held transactions (which implies there are multiple transactionIDs that need to be returned).

This means that the `childHoldData` object must be returned elsewhere and not at the current location where the `childCanHold` and `childCanUnhold` flags are.

Let me know if that makes sense- could you please link your draft PR so I can see where this is happening? Thanks!

**Comment by cdOut** [comment]
Hm the preview report action seems like the most suitable place to include this sort of data as it's very easy to read in the context menu code on the frontend, other places might prove to be problematic.

I'll raise the draft PR for it in a moment and cc you so you can take a look.

**Comment by cdOut** [comment]
@robertjchen [Here's](https://github.com/Expensify/App/pull/51354) the draft for it, I've prepared it fully to support the suggested Hold data structure.

**Comment by robertjchen** [comment]
@cdOut I see, but the problem remains that for the report preview, we can have child reports with multiple transactions are either on hold or not on hold.

Maybe the best approach is to just return the first held transaction in `childHoldData` (even if there are multiple held transactions) and run with that for now?

**Comment by cdOut** [comment]
Well if the preview is a summary of multiple expenses, we should pass `null / undefined` as `childHoldData`. Only if it's a reportAction of a preview for a singular expense we should pass the required data.

**Comment by robertjchen** [comment]
👍

**Comment by robertjchen** [comment]
That makes sense, I'll look into that now

**Comment by robertjchen** [comment]
This also means that in the case where a child transaction is on hold, but cannot be unheld (due to permissions), we will return `null / undefined` as `childHoldData`.

**Comment by cdOut** [comment]
Yup, that's exactly it.

**Comment by cdOut** [comment]
@robertjchen any updates on the backend for this issue? Thanks!

**Comment by robertjchen** [comment]
Testing out some local changes at the moment, aiming on having the changes reviewed and deployed tomorrow

**Comment by robertjchen** [comment]
Draft PR for BE changes up shortly

**Comment by blazejkustra** [comment]
@robertjchen Please ping me once that's deployed, thanks! 🙌

**Comment by sumo-slonik** [comment]
Hi everyone, unfortunately @cdOut  has ended his cooperation with us and now I will take care of the completion of this issue. I will be very grateful for updates as soon as the backend is ready.

**Comment by robertjchen** [comment]
Circling back to this just now, will update shortly

**Comment by sumo-slonik** [comment]
@robertjchen Hi, what is the status regarding this issue?

**Comment by robertjchen** [comment]
no update

**Comment by mallenexpensify** [comment]
Bumping to Daily and moved to HIGH status (ie. UX Reliability, API Reliability and Performance issues that are part of main flows and affect a subset of users. ) 

Assigned to @sumo-slonik , also reassigned @jliexpensify cuz we want a BZ on all bug issues (and you were previously assigned)
@sumo-slonik and @jliexpensify can you check E/App issues to see if this is a dupe? I've seen many held/hold issues lately. 

Mainly looking for a substantive update here so we know where we're at and so we can prioritize accordingly.  Thx

**Comment by robertjchen** [comment]
Yes, this was deployed back Oct 2024, there should be two new fields in the parent report summary `childCanHold` and `childCanUnhold`

**Comment by sumo-slonik** [comment]
Thanks for update @robertjchen.
I am now busy implementing a design doc related to a new simple report flow with @mountiny. If you think this has a higher priority I can change the context to this.

**Comment by mountiny** [comment]
Can this be exported to the community? @robertjchen

**Comment by robertjchen** [comment]
Yep, given that the fields are in place, someone could try having a crack at this one! 🙏

**Comment by mountiny** [comment]
@robertjchen Feel free to export the issue for the community next time!

**Comment by CyberAndrii** [comment]
I think `childCanHold` and `childCanUnhold` were added in the wrong place. They should be on individual transaction actions (where `action.actionName === 'iou'`), rather than on the parent report action. But then they should be renamed to `canHold` and `canUnhold`. That's how it's done on the search page.

https://github.com/Expensify/App/blob/52b22cb56e98a3e216186432f26ea7d201297c23/src/types/onyx/SearchResults.ts#L302-L306

**Comment by CyberAndrii** [comment]
> Eh, I didn't find these two fields in the response. 😅

It's on the `REPORTPREVIEW` action

<details>
 <summary>Image</summary>

<img src="https://github.com/user-attachments/assets/a960286c-7a0d-4817-91ae-49e02819fa2a">

</details>

But we also need them on `IOU` actions

<details>
 <summary>Image</summary>

<img src="https://github.com/user-attachments/assets/bb266c31-8a92-45e4-b887-f00d9f55b238">

</details>

And it should be null or false on `REPORTPREVIEW` actions that contain multiple expenses. https://github.com/Expensify/App/issues/47241#issuecomment-2435028452

**Comment by abzokhattab** [comment]
## Proposal 


### Please re-state the problem that we are trying to solve in this issue. 


The "Hold" button on the expense preview appears only if the expense is opened first.


### What is the root cause of that problem? 


The "Hold" button on the expense preview does not appear because the necessary child data to determine if an expense can be held is only available after fully opening the expense, rather than being included in the preview action.


### What changes do you think we should make in order to solve the problem? 

 
- Inside the [canHoldUnholdReportAction](https://github.com/Expensify/App/blob/0269e95a3c842b8c5722489c5e676ddd1e41d2d6/src/libs/ReportUtils.ts#L3745) , we should add a check to see if the parent report action is a preview and if it contains the newly added parameters:



```js
if (isReportPreviewAction(parentReportAction) && parentReportAction.childCanHold !== undefined && parentReportAction.childCanUnhold !== undefined) {
        return {
            canHoldRequest: parentReportAction.childCanHold,
            canUnholdRequest: parentReportAction.childCanUnhold,
        };
    }
```

 
- We also need to update the condition here to pass the `reportAction` as a fallback to `canHoldUnholdReportAction` when the `moneyRequestAction` is `undefined`. This can be done [here](https://github.com/Expensify/App/blob/3dffdc2cc98f7590bdab8dfa406c7b48785cc81b/src/pages/home/report/ContextMenu/ContextMenuActions.tsx#L357)  and [here](https://github.com/Expensify/App/blob/3dffdc2cc98f7590bdab8dfa406c7b48785cc81b/src/pages/home/report/ContextMenu/ContextMenuActions.tsx#L374)  as follows:



```js
type === CONST.CONTEXT_MENU_TYPES.REPORT_ACTION && areHoldRequirementsMet && canHoldUnholdReportAction(moneyRequestAction ?? reportAction).canUnholdRequest,
...
type === CONST.CONTEXT_MENU_TYPES.REPORT_ACTION && areHoldRequirementsMet && canHoldUnholdReportAction(moneyRequestAction ?? reportAction).canHoldRequest,
```

 
- Also, update the [areHoldRequirementsMet](https://github.com/Expensify/App/blob/b1cd8e2cd36a7bfeb53a96b5fa74275bf62d7992/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L199)  function to return `true` if `moneyRequestAction` is undefined and `reportAction` exists, as this is the only case where the condition will be true:



```js
const areHoldRequirementsMet =
        (!isInvoiceReport &&
            isMoneyRequestOrReport &&
            !isArchivedNonExpenseReport(transactionThreadReportID ? childReport : parentReport, transactionThreadReportID ? childReportNameValuePairs : parentReportNameValuePairs)) ||
        (!!reportAction && !moneyRequestAction);
```


Result: 


https://github.com/user-attachments/assets/a89a638d-5cc0-4037-86d8-8f173d978920



also i notice a problem where the backend doenst update the `childCanHold` and `childCanUnhold` whenever the request is changed to held or unheld
### What specific scenarios should we cover in automated tests to prevent reintroducing this issue in the future? 


NA


### What alternative solutions did you explore? (Optional) 


NA

**Comment by jliexpensify** [comment]
👋 Hi, how are we going with this issue @ntdiary ?

**Comment by jliexpensify** [comment]
Got it, thanks for the update! cc @robertjchen for this [comment](https://github.com/Expensify/App/issues/47241#issuecomment-2769827004)

**Comment by robertjchen** [comment]
this week

**Comment by mallenexpensify** [comment]
@hoangzinh plz take over as C+. thx

**Comment by jliexpensify** [comment]
Still waiting on @robertjchen right?

**Comment by jliexpensify** [comment]
@hoangzinh let me check in with @robertjchen via DM around this. Thanks!

**Comment by robertjchen** [comment]
Given current priorities I think we can de-prioritize this for now. In the meantime, I will ask around internally to see if anyone else is available to pick this up 👍

**Comment by robertjchen** [comment]
no update

**Comment by LorenzoBloedow** [comment]
🚨 Edited by **proposal-police**: This proposal was **edited** at 2025-06-12 14:04:37 UTC.

## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The hold button from the context menu on report previews only appears after opening the report.

### What is the root cause of that problem?
We're not correctly finding the IOU report for the report preview.

### What changes do you think we should make in order to solve the problem?
1. Find all the IOUs in the report actions, if any.
```ts
   const childReportMoneyRequestActions = useMemo(() => {
        if (isMoneyRequestReport || isInvoiceReport) {
            if (!paginatedReportActions) {
                return [parentReportAction]
            }
            return paginatedReportActions.filter((action) => action.actionName === CONST.REPORT.ACTIONS.TYPE.IOU && !isDeletedAction(action));
        }
        return [parentReportAction];
    }, [parentReportAction, isMoneyRequestReport, isInvoiceReport, paginatedReportActions]);

    const moneyRequestAction = childReportMoneyRequestActions.at(0);
```
Instead of:
https://github.com/Expensify/App/blob/e1766ef329607a44d08fe805e22e11d26fab6f25/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L181-L191

2. Make sure we don't show the hold/unhold buttons when there are multiple expenses because *currently* report previews don't seem to able to tell which expense we're clicking on:
```ts
    const areHoldRequirementsMet =
        !isInvoiceReport &&
        isMoneyRequestOrReport &&
        !isArchivedNonExpenseReport(transactionThreadReportID ? childReport : parentReport, transactionThreadReportID ? childReportNameValuePairs : parentReportNameValuePairs) &&
        childReportMoneyRequestActions.length < 2;
```
Instead of:
https://github.com/Expensify/App/blob/e1766ef329607a44d08fe805e22e11d26fab6f25/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L202-L205

### What specific scenarios should we cover in automated tests to prevent reintroducing this issue in the future?
N/A (Not sure if we can test `BaseReportActionContextMenu` with a UI test)

### What alternative solutions did you explore? (Optional)
N/A

Result (One transaction):

https://github.com/user-attachments/assets/36af00a9-c0db-46a5-b609-36300d3f3434

Result (Multiple transactions):

https://github.com/user-attachments/assets/ecf6025e-dc85-4fe4-88a1-e45d80df96d7

**Comment by hoangzinh** [comment]
Thank you for posting proposal @LorenzoBloedow. I will try to review it soon

**Comment by hoangzinh** [comment]
Sorry, I need time to look through all comments and understand previous concerns here.

**Comment by hoangzinh** [comment]
@LorenzoBloedow I know this [comment](https://github.com/Expensify/App/issues/47241#issuecomment-2769827004) is old, there have been a lot of changes since then, especially the new report review. Could you please review that comment and clarify whether your proposal has already addressed these points?

**Comment by LorenzoBloedow** [comment]
@hoangzinh I was reviewing that [comment](https://github.com/Expensify/App/issues/47241#issuecomment-2769827004) and my proposal and I just noticed the backend is already sending all the data necessary to determine if we can hold/unhold, and that my proposal was getting `childCanUnhold/hold` from the wrong action, which means all the videos are the result of calling the `canHoldUnholdReportAction` function without ever considering `childCanUnhold/hold`.

So I tried using the correct `childCanUnhold/hold` and the logic breaks because the issues in this [comment](https://github.com/Expensify/App/issues/47241#issuecomment-2769827004) don't seem to have been addressed yet.

This might be a weird request, but can we just ditch the `childCanHold/childCanUnhold` parameters for this specific context? Seems like we already have all the data we need and the backend logic is not fully finished yet. I'm sure it could be used in other contexts where the necessary data isn't available though.

I updated my [proposal](https://github.com/Expensify/App/issues/47241#issuecomment-2953369908) to show how it works without the new parameters.

Also, I tested it, and the expected behavior described in that [comment](https://github.com/Expensify/App/issues/47241#issuecomment-2769827004) seems to have been addressed already by the FE, including not showing the hold button to the submitter when an admin holds:

https://github.com/user-attachments/assets/b88a7cbe-48f9-4185-9b40-ddfb15ea1952

**Comment by hoangzinh** [comment]
I got your point. Let me check if we can just find the correct iou report action in this case, and will it cause any regression, especially in the delete action.

**Comment by hoangzinh** [comment]
@LorenzoBloedow can you clear Onyx or sign out/sign in again and test if your solution still works? From my investigation, it seems `paginatedReportActions` doesn't contain any iou action until we open the thread. Thus your suggestion _"Find all the IOUs in the report actions, if any"_ will always return null

**Comment by LorenzoBloedow** [comment]
@hoangzinh Sorry, I didn't think to clear Onyx when testing, so you're right, I've tested it again and the backend is not sending the IOU action with the report.

It seems we have 3 options I could think of:
* Wait for BE to fix the issues with `childCanHold` and `childCanUnhold`.
* Ask for the BE to send IOU action instead of the new fields (In my opinion, this is the best option because we already have [optimistic logic](https://github.com/Expensify/App/blob/a7e43b570e3fa9863a189c08bf1c62a95c4dcea6/src/libs/ReportUtils.ts#L4031-L4079) for showing hold/unhold, not sure how that would work with the new fields).
* Fetch the IOU actions when the context menu is opened/report preview is hovered (this is the worst option IMO, but fastest to implement).

PS: An interesting behavior I've noticed is that while the App is open we do get the IOU action from Pusher. That doesn't happen when populating the cache for the first time though:

https://github.com/user-attachments/assets/65d78ff0-d4f7-4417-9760-f5a072eb2532

**Comment by hoangzinh** [comment]
Thank you @LorenzoBloedow 

@robertjchen, which of the above options do you prefer? I understand that we preferred Option 1, `Wait for BE to fix the issues with childCanHold and childCanUnhold` but wanna clarify again.

**Comment by robertjchen** [comment]
👀

**Comment by jliexpensify** [comment]
Waiting on @robertjchen

**Comment by hoangzinh** [comment]
Same as above. @robertjchen do you have any chance to look at this issue?

**Comment by robertjchen** [comment]
👀

**Comment by robertjchen** [comment]
checking this today
